### PR TITLE
Scheduler V2 Fixes

### PR DESCRIPTION
**Pull request by lmars**
The main change here is making sure the scheduler is in sync with hosts and formations before starting the main loop.

I have also changed the startup so that the scheduler only registers in service discovery once it is in sync
with the hosts so that a deployment doesn't kill old schedulers before the new schedulers are watching events, which could lead to events being missed. This meant doing discoverd related things in `Run`, so I moved the functionality into an interface to avoid running a discoverd server in the tests.

### CODE REVIEW

File: controller/scheduler-v2/scheduler.go [hunk 1]
```diff
@@ -191,15 +139,29 @@ func (s *Scheduler) streamFormationEvents() error {
...
 	go func() {
+		var isCurrent bool
 	outer:
 		for {
 			for formation := range events {
+				// an empty formation indicates we now have the
+				// current list of formations.
 				if formation.App == nil {
-					// sentinel
+					if !isCurrent {
```
Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: Why is this necessary for the formation list? The scheduler won't try to stop anything it doesn't have a formation for.

Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: nvmind, probably better to have this than not.

File: controller/scheduler-v2/scheduler.go [hunk 2]
```diff
@@ -568,33 +554,41 @@ func (s *Scheduler) sendDiffRequests(f *Formation, diff Processes) {
...
+	log.Info("getting active jobs")
+	jobs, err := h.ListJobs()
+	if err != nil {
+		log.Error("error getting active jobs", "err", err)
+		return
+	}
+	log.Info(fmt.Sprintf("got %d active job(s) for host %s", len(jobs), h.ID()))
+
+	for _, job := range jobs {
+		s.handleActiveJob(&job)
```
Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: This will happen already as part of StreamEvents.

**Comment by jzila** [comment]
LGTM other than my comment on calling `handleActiveJob` in `followHost`

File: controller/scheduler-v2/scheduler.go (hunk 2 above)
Comment (lmars) [inline controller/scheduler-v2/scheduler.go]: `StreamEvents` does not return the current state like other streams do, so this makes sure we have the current state when we start following a host (rather than relying on the next trigger of `SyncJobs`).

File: controller/scheduler-v2/scheduler.go (hunk 1 above)
Comment (lmars) [inline controller/scheduler-v2/scheduler.go]: I think ensuring the scheduler is in sync before taking (or not taking) any action is a good idea.
